### PR TITLE
[Simt] Add arch check for shared array.

### PR DESCRIPTION
**Pull request by turbo0628**
Shared array didn't check arch. 

Raise an exception early so that users can get clear error information.

### CODE REVIEW

File: tests/python/test_shared_array.py [hunk 1]
```diff
@@ -46,9 +47,18 @@ def calc_shared_array(v: ti.types.ndarray(ndim=1),
     assert np.allclose(reference, a_arr)
 
 
-@test_utils.test(arch=[ti.cuda, ti.vulkan])
+@test_utils.test(arch=[ti.cpu, ti.cuda, ti.vulkan])
 def test_shared_array_nested_loop():
-    _test_shared_array_nested_loop()
+    arch = ti.lang.impl.get_runtime().prog.config().arch
```
Comment (ailzhang) [inline tests/python/test_shared_array.py:52]: aha better split cuda/vulkan & cpu into different tests so that it's clear to expect throws in cpu.

**Review by ailzhang — APPROVED** [review]
LGTM!

**Comment by taichi-gardener** [comment]
Rebase failed, please resolve the conflicts manually.

**Comment by taichi-gardener** [comment]
Rebase failed, please resolve the conflicts manually.

**Comment by bobcao3** [comment]
Two questions, 1. are we still doing this? 2. I think i recently enabled shared array for opengl, metal, and dx11, and I think some gate keeping code are already there?

**Pull request closed without merging**